src/chainstate/stacks/boot/pox.clar

```diff
@@ -380,6 +458,141 @@
+  (let ((partial-stacked
+         ;; fetch the partial commitments
+         (unwrap! (map-get? partial-stacked-by-cycle { pox-addr: pox-addr, sender: tx-sender, reward-cycle: reward-cycle })
+                  (err ERR_STACKING_NO_SUCH_PRINCIPAL))))
+    ;; must be called directly by the tx-sender or by an allowed contract-caller
+    (asserts! (check-caller-allowed)
+              (err ERR_STACKING_PERMISSION_DENIED))
+    (let ((amount-ustx (get stacked-amount partial-stacked)))
+      (try! (can-stack-stx pox-addr amount-ustx reward-cycle u1))
+      ;; add the pox addr to the reward cycle
+      (add-pox-addr-to-ith-reward-cycle
+       u0
+       { pox-addr: pox-addr,
+         first-reward-cycle: reward-cycle,
+         num-cycles: u1,
+         amount-ustx: amount-ustx,
+         i: u0 })
+      ;; don't update the stacking-state map,
+      ;;  because it _already has_ this stacker's state
+      ;; don't lock the STX, because the STX is already locked
+      ;;
+      ;; clear the partial-stacked state
+      (map-delete partial-stacked-by-cycle { pox-addr: pox-addr, sender: tx-sender, reward-cycle: reward-cycle })
+      (ok true))))
+
+;; As a delegate, stack the given principal's STX using partial-stacked-by-cycle
+;; Once the delegate has stacked > minimum, the delegate should call stack-aggregation-commit
+(define-public (delegate-stack-stx (stacker principal)
+                                   (amount-ustx uint)
+                                   (pox-addr { version: (buff 1), hashbytes: (buff 20) })
+                                   (lock-period uint))
+    ;; this stacker's first reward cycle is the _next_ reward cycle
+    (let ((first-reward-cycle (+ u1 (current-pox-reward-cycle)))
+          (unlock-burn-height (reward-cycle-to-burn-height (+ (current-pox-reward-cycle) u1 lock-period))))
+
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+        (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; stacker must have delegated to the caller
+      (asserts! 
+        (let ((delegation-info (unwrap! (get-check-delegation stacker) (err ERR_STACKING_PERMISSION_DENIED))))
+               ;; must have delegated to tx-sender
+          (and (is-eq (get delegated-to delegation-info) tx-sender)
+               ;; must have delegated enough stx
+               (>= (get amount-ustx delegation-info) amount-ustx)
+               ;; if pox-addr is set, must be equal to pox-addr
+               (match (get pox-addr delegation-info)
+                      specified-pox-addr (is-eq pox-addr specified-pox-addr)
+                      true)
+               ;; delegation must not expire before lock period
+               (match (get until-burn-ht delegation-info)
+                      until-burn-ht (>= until-burn-ht
+                                        unlock-burn-height)
+                      true)))
+        (err ERR_STACKING_PERMISSION_DENIED))
```

Comment:
1 assert, 1 test, 1 error would probably increase readability, but also the UX for the people interacting with that contract.

Reply:
Yeah, good idea -- will add a ERR_STACKING_PERMISSION_EXPIRED

Reply:
Updated in 4745049f2, added:
```
(define-constant ERR_DELEGATION_EXPIRES_DURING_LOCK 21)
(define-constant ERR_DELEGATION_TOO_MUCH_LOCKED 22)
(define-constant ERR_DELEGATION_POX_ADDR_REQUIRED 23)
```